### ISSUE TITLE
macOS dings when key is pressed

### ISSUE THREAD

**Issue by treeplate** [flutter/flutter#74287]
## Steps to Reproduce

1. Run `flutter create ding_bug`.
2. Update the files as follows: ... (this is the sample in the `LogicalKeyboardKey` docs)
<details>
<summary>main.dart</summary>

```
/// Flutter code sample for LogicalKeyboardKey

// This example shows how to detect if the user has selected the logical "Q"
// key.

import 'package:flutter/material.dart';

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

void main() => runApp(MyApp());

/// This is the main application widget.
class MyApp extends StatelessWidget {
  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: _title,
      home: Scaffold(
        appBar: AppBar(title: const Text(_title)),
        body: MyStatefulWidget(),
      ),
    );
  }
}

/// This is the stateful widget that the main application instantiates.
class MyStatefulWidget extends StatefulWidget {
  MyStatefulWidget({Key key}) : super(key: key);

  @override
  _MyStatefulWidgetState createState() => _MyStatefulWidgetState();
}

/// This is the private State class that goes with MyStatefulWidget.
class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  // The node used to request the keyboard focus.
  final FocusNode _focusNode = FocusNode();
// The message to display.
  String _message;

// Focus nodes need to be disposed.
  @override
  void dispose() {
    _focusNode.dispose();
    super.dispose();
  }

// Handles the key events from the RawKeyboardListener and update the
// _message.
  void _handleKeyEvent(RawKeyEvent event) {
    setState(() {
      if (event.logicalKey == LogicalKeyboardKey.keyQ) {
        _message = 'Pressed the "Q" key!';
      } else {
        if (kReleaseMode) {
          _message =
              'Not a Q: Key label is "${event.logicalKey.keyLabel ?? '<none>'}"';
        } else {
          // This will only print useful information in debug mode.
          _message = 'Not a Q: Pressed ${event.logicalKey.debugName}';
        }
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    final TextTheme textTheme = Theme.of(context).textTheme;
    return Container(
      color: Colors.white,
      alignment: Alignment.center,
      child: DefaultTextStyle(
        style: textTheme.headline4,
        child: RawKeyboardListener(
          focusNode: _focusNode,
          onKey: _handleKeyEvent,
          child: AnimatedBuilder(
            animation: _focusNode,
            builder: (BuildContext context, Widget child) {
              if (!_focusNode.hasFocus) {
                return GestureDetector(
                  onTap: () {
                    FocusScope.of(context).requestFocus(_focusNode);
                  },
                  child: const Text('Tap to focus'),
                );
              }
              return Text(_message ?? 'Press a key');
            },
          ),
        ),
      ),
    );
  }
}
```
</details>
3. Tap and press 'Q'

**Expected results:** it saying 'Press the Q key'

**Actual results:** same as expected, but with a alert sound on MacOs

**Comment by pedromassangocode** [comment]
I was able to reproduce.  This sounds more like a proposal since the current API is expected to work as it is.

<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel master, 1.26.0-2.0.pre.448, on Mac OS X 10.15.7 19H114 darwin-x64, locale en)
    • Flutter version 1.26.0-2.0.pre.448 at /Users/pedromassango/Code/flutter_master
    • Framework revision eeb7c261f2 (3 hours ago), 2021-01-20 23:04:03 -0500
    • Engine revision dde6b7e8ea
    • Dart version 2.12.0 (build 2.12.0-246.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Community Edition (version 2020.3.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 52.2.5
    • Dart plugin version 203.6912

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • Mac OS X 10.15.7 19H114 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 87.0.4280.141

! Doctor found issues in 1 category.
```

</details>

**Comment by stuartmorgan** [comment]
/cc @gspencergoog 

I forgot when we implemented passing keyDown events to the next responder that by default keyDowns that propagate all the way to the top of the hierarchy without being handled will beep. We don't want keyDown to be a black hole for add-to-app scenarios though.

Options:
- Require Flutter apps that don't want to beep to claim to handle all keyDown events
- Make the default behavior to not pass keyDown/keyUp (but still performKeyEquivalent:; I think that will work), and have an API that can be used to enable event propagation for add-to-app.

**Comment by Hixie** [comment]
If there's any way to enable developers (even in non-add-to-app scenarios) to say "I can't handle this" about a key, and have the system beep, that would be ideal. Failing that, it's ok if we just have a way to manually do the beep (honoring all the relevant settings in the sounds control panel, of course). That's basically what we do with haptic feedback on mobile platforms -- we have an API to reproduce the effect.

**Comment by stuartmorgan** [comment]
> If there's any way to enable developers (even in non-add-to-app scenarios) to say "I can't handle this" about a key, and have the system beep, that would be ideal.

There is: the (as of recently) current behavior. The description of this issues was for some reason reversed during triage; I've now restored the original.

The problem is that if a view accepts focus (as `FlutterView` does), any keyDown that the view (i.e., the application, from the perspective of a non-add-to-app Flutter developer) doesn't explicitly say it handles will beep. We need to decide whether we actually want that behavior.

**Comment by Hixie** [comment]
Seems like reasonable behaviour if we document how to say that we did handle it. On macOS that seems to be the way things are intended to work, right?

**Comment by stuartmorgan** [comment]
> On macOS that seems to be the way things are intended to work, right?

I thought it was only for views that declared that they accepted focus, but I tested a hello-world macOS app and it does indeed beep on every keypress with just a stock `NSView`, so yes, this is WAI. Based on that, closing.

@treeplate If you want to avoid beeping for keys you should use the `FocusNode`'s `onKey`, to actually handle the event (i.e., return `true`), rather than just listening for the existence of the event.

**Comment by gspencergoog** [comment]
@treeplate Handling the key with the `FocusNode.onKey` is easiest when done with a `Focus` widget:

```
Widget build(BuildContext context) {
  return Focus(
    onKey: (FocusNode node, RawKeyEvent event) => true,
	child: ...
  );
}
```

That will manage the focus node for you (inserting and removing it as necessary).

**Comment by Hixie** [comment]
We should update the docs for `LogicalKeyboardKey` to answer that, since if I understand the OP correctly they literally just copied that example.

**Comment by Hixie** [comment]
(Reopening per my last comment.)

**Comment by mykdavies** [comment]
This issue also seems to prevent use of `KeyboardListener` and `RawKeyboardListener` when targetting macOS, as their `onKey:` argument takes `void Function(FocusNode, RawKeyEvent)?` rather than `KeyEventResult Function(FocusNode, RawKeyEvent)?` and so cannot return `KeyEventResult.handled`.

**Comment by gspencergoog** [comment]
No, it doesn't prevent them from being used, but that's also not their intended purpose. The widgets `KeyboardListener` and `RawKeyboardListener` are purely listeners: they don't handle keys, they just tell you about the events.

If you want to handle keys, you should use a `Focus` widget with an `onKey` handler, as I mentioned above, although with the current version of the API, instead of returning `true`, you'd return `KeyEventResult.handled`. If all you want to do is prevent the "ding", you can place the `Focus` widget above the `MaterialApp` or `CupertinoApp` in you application (basically, very near the `runApp`, above any other handlers).

**Comment by mykdavies** [comment]
Yes, I am now using a Focus widget in that way thanks to the discussion above.

I'd not really registered that they are "Listeners" rather than "Handlers", thanks for clearing that up. It's unfortunate that the `KeyboardListener` docs don't call that out and point readers to the very comprehensive `Focus` docs.

**Comment by mykdavies** [comment]
That's excellent news, thanks for your replies here too!

**Comment by cjng96** [comment]
Using Focus widget to prevent a beep sound is not suitable solution if we are using text input like TextFiled widget.
because TextField widget it not working if we return KeyEventResult.handled in onKeyEvent callback.

If you are using text input, such as the TextFiled widget, using the Focus widget to prevent beeps is not an appropriate solution.
This is because the TextField widget does not work when KeyEventResult.handled is returned in onKeyEvent callback.

Is there any other solution?

**Comment by gspencergoog** [comment]
@cjng96 Please file a new issue for the `TextField` issue you are describing so that we can triage it separately, ideally with a small sample app that reproduces the problem.

**Comment by lucas-goldner** [comment]
> @treeplate Handling the key with the `FocusNode.onKey` is easiest when done with a `Focus` widget:
> 
> ```
> Widget build(BuildContext context) {
>   return Focus(
>     onKey: (FocusNode node, RawKeyEvent event) => true,
> 	child: ...
>   );
> }
> ```
> 
> That will manage the focus node for you (inserting and removing it as necessary).

Hey, this has been open for a while. I guess this has been forgotten a bit. 
I am replying to the comment and updating the solution since the API has changed.
You have to return KeyEventResult now instead of a boolean.

```dart
Focus(
    onKey: (focus, onKey) => KeyEventResult.handled,
)
```

**Comment by crcdng** [comment]
The proposed solution handling the key with the FocusNode.onKey with a Focus widget does not work. The system sound still occurs.

**Comment by sandapilarius** [comment]
fyi,

class MutedWidgetsFlutterBinding extends WidgetsFlutterBinding {
  static WidgetsBinding ensureInitialized() {
    MutedWidgetsFlutterBinding();
    return WidgetsBinding.instance;
  }

  @override
  BinaryMessenger createBinaryMessenger() {
    return SystemSoundMuterBinaryMessenger(super.createBinaryMessenger());
  }
}

class SystemSoundMuterBinaryMessenger extends BinaryMessenger {
  final BinaryMessenger _delegate;

  SystemSoundMuterBinaryMessenger(this._delegate);

  @override
  Future<ByteData?>? send(String channel, ByteData? message) {
    if (channel == 'flutter/platform' && message != null) {
      try {
        final MethodCall call = const JSONMethodCodec().decodeMethodCall(message);
        if (call.method == 'SystemSound.play') {
          return Future.value(const JSONMethodCodec().encodeSuccessEnvelope(null));
        }
      } catch (_) {}
    }
    return _delegate.send(channel, message);
  }

  @override
  void setMessageHandler(String channel, MessageHandler? handler) {
    _delegate.setMessageHandler(channel, handler);
  }

  @override
  Future<void> handlePlatformMessage(
    String channel,
    ByteData? message,
    PlatformMessageResponseCallback? callback,
  ) {
    // ignore: deprecated_member_use
    return _delegate.handlePlatformMessage(channel, message, callback);
  }
}
